lib_eio/sync.ml

```diff
@@ -0,0 +1,448 @@
+   the producer's value is passed to the consumer's callback. If the consumer accepts it,
+   then both fibers are resumed. If not, the producer starts again (incrementing [balance]
+   again) and waits for another consumer.
+
+   The above has not been formally verified (exercise for reader!). *)
+
+(* Import these directly because we copy this file for the dscheck tests. *)
+module Fiber_context = Eio__core.Private.Fiber_context
+module Suspend = Eio__core.Private.Suspend
+module Cancel = Eio__core.Cancel
+
+type 'a item = {
+  v : 'a;
+  kp : (bool, exn) result -> unit;      (* [Ok false] means consumer refused the item; retry. *)
+  cancel : [
+    | `Resuming                         (* In the process of resuming, so can't cancel. *)
+    | `Suspended of (unit -> bool)      (* Call this function to attempt to leave the queue. *)
+    | `Cancelled of exn                 (* Already cancelled. *)
+  ] Atomic.t;
+}
+
+type 'a cell =
+  | In_transition
+  | Slot of ('a -> bool)
+  | Item of 'a item
+  | Finished
+
+module Cell = struct
+  type 'a t = 'a cell
+
+  let init = In_transition
+
+  let segment_order = 2
+
+  let dump f = function
+    | In_transition -> Fmt.string f "In_transition"
+    | Slot _ -> Fmt.string f "Slot"
+    | Item _ -> Fmt.string f "Item"
+    | Finished -> Fmt.string f "Finished"
+end
+
+module Q = Cells.Make(Cell)
+
+type 'a t = {
+  balance : int Atomic.t;
+  consumers : 'a Q.t;
+  producers : 'a Q.t;
+}
+
+type 'a loc =
+  | Short of 'a Cell.t Atomic.t                 (* Acting as resumer of cell *)
+  | Long of ('a Q.segment * 'a Cell.t Atomic.t) (* Acting as suspender of cell; can cancel *)
+
+let dump f t =
+  Fmt.pf f "@[<v2>Sync (balance=%d)@,@[<v2>Consumers:@,%a@]@,@[<v2>Producers:@,%a@]@]"
+    (Atomic.get t.balance)
+    Q.dump t.consumers
+    Q.dump t.producers
+
+(* Give [item] to consumer [kc]. [item]'s cell is now Finished. *)
+let exchange item kc = item.kp (Ok (kc item.v))
+
+(* Add [value] to [cell].
+   If the cell is in transition, place [value] there and let the other party handle it later.
+   If the peer's value is already present, do the exchange.
+   If the peer cancelled the cell then try the next one on the given resume queue (if we're adding
+   to a suspend queue then it can't be cancelled, because the caller controls cancellation).
+   This is only used when our fiber is already suspended,
+   since we can't create [value] before we have the continuation. *)
+let rec add_to_cell queue value cell =
+  match Atomic.get cell, value with
+  | Finished, _ -> add_to_cell queue value (Q.next_resume queue)     (* Cancelled - skip *)
+  | (Slot kc   as old), Item item
+  | (Item item as old), Slot kc ->
+    if Atomic.compare_and_set cell old Finished then exchange item kc
+    else add_to_cell queue value cell
+  | In_transition, _ ->
+    if Atomic.compare_and_set cell In_transition value then ()
+    else add_to_cell queue value cell
+  | (Slot _ | Item _), _ -> assert false
```

Comment:
Just a thought: You could probably rule this case out by using either GADTs or polymorphic variants and make the two queues to be of different type.  I'm not sure it is worth the effort, but it should be possible.

Reply:
That doesn't work - you can get either type on either queue (it just depends who writes the cell first). e.g. on the `consumers` queue usually the consumer gets there first and writes a `Slot`, sometimes the producer gets there first and writes an `Item` instead.